src/core/Oscillator.cpp

```diff
@@ -182,19 +182,20 @@ void Oscillator::generateFromFFT(int bands, sample_t* table)
 	normalize(s_sampleBuffer.data(), table, OscillatorConstants::WAVETABLE_LENGTH, 2*OscillatorConstants::WAVETABLE_LENGTH + 1);
 }
 
-void Oscillator::generateAntiAliasUserWaveTable(SampleBuffer *sampleBuffer)
+std::unique_ptr<OscillatorConstants::waveform_t> Oscillator::generateAntiAliasUserWaveTable(const SampleBuffer* sampleBuffer)
 {
-	if (sampleBuffer->m_userAntiAliasWaveTable == nullptr) {return;}
-
+	auto userAntiAliasWaveTable = std::make_unique<OscillatorConstants::waveform_t>();
 	for (int i = 0; i < OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT; ++i)
 	{
 		for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH; ++i)
 		{
-			s_sampleBuffer[i] = sampleBuffer->userWaveSample((float)i / (float)OscillatorConstants::WAVETABLE_LENGTH);
+			s_sampleBuffer[i] = Oscillator::userWaveSample(sampleBuffer, (float)i / (float)OscillatorConstants::WAVETABLE_LENGTH);
```

Comment:
So the code seems to be populating the entire waveform ``userAntiAliasWaveTable`` by first creating a wave table (i.e., calling ``userWaveSample`` throughout the entire buffer), and then putting that wave table in the wave form. It then does this until the entire waveform is populated. If we had removed the inner-loop and just said ``s_sampleBuffer[i] = Oscillator::userWaveSample(sampleBuffer, (float)i / (float)OscillatorConstants::WAVETABLE_LENGTH);``, then that wouldn't populate the wave table correctly.

I do think though, that it is probably inefficient to keep calling ``userWaveSample`` after the first wave table has been made, as you are right that it does seem to be doing the same thing, with the code looking like it should generate identical wave tables.

To keep it simple though, I'll just change the variable name of the inner loop so it isn't confused with the outer loop, and replace the C-style cast with ``static_cast``.

Reply:
Yeah, we can't remove the inner loop. I meant that it should be moved outside the loop its in. So instead of M x N iterations, it would be M + N.

Maybe you could put a TODO comment there as a reminder to fix it in a later PR?